src/Interpreters/TLSLog.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <Interpreters/SystemLog.h>
+#include <Core/NamesAndTypes.h>
+#include <Core/NamesAndAliases.h>
+#include <Columns/IColumn.h>
+#include <Storages/ColumnsDescription.h>
+
+namespace Poco
+{
+namespace Net
+{
+class Socket;
+class X509Certificate;
+class VerificationErrorArgs;
+}
+}
+
+namespace DB
+{
+
+
+constexpr int64_t TLSLogErrorFlushIntervalSeconds = 10;
+enum class TLSLogElementType : int8_t
+{
+    SESSION_FAILURE = 0,
+    SESSION_SUCCESS = 1,
+};
+
+/**
+A struct which will be inserted as row into tls_log table.
+*/
+struct TLSLogElement
+{
+    using Type = TLSLogElementType;
+
+    explicit TLSLogElement(const String &failure_reason = "", uint64_t failure_count = 0);
+#if USE_SSL
+    explicit TLSLogElement(const Poco::Net::X509Certificate &tls_certificate, const String &user, const String &failure_reason = "", uint64_t failure_count = 0);
+#endif
+
+    TLSLogElementType type;
+    time_t event_time{};
+    Decimal64 event_time_microseconds{};
+    Strings certificate_subjects;
+    std::pair<time_t, time_t> time_range;
+    String certificate_serial;
+    String certificate_issuer;
+    String user;
+    String failure_reason;
+    uint64_t failure_count = 0;
+
+    static std::string name() { return "TLSLog"; }
+
+    static NamesAndTypesList getNamesAndTypes();
+    static NamesAndAliases getNamesAndAliases() { return {}; }
+    static ColumnsDescription getColumnsDescription();
+    static const char * getCustomColumnList() { return nullptr; }
+
+    void appendToBlock(MutableColumns & columns) const;
+};
+
+/** TLSLog contains information about client certificates received
+by ClickHouse server during TLS handshakes.*/
+class TLSLog : public SystemLog<TLSLogElement>
+{
+    using TimePoint = std::chrono::system_clock::time_point;
+    struct CertStats
+    {
+        int64_t count;
+        std::unique_ptr<Poco::Net::X509Certificate> certificate;
+        String last_error_message;
```

Comment:
It make sense to store the time if the last error and use it when the cache of errors is flushed.
Otherwise all the error have the same `event_time_microseconds` which equals to time of the flush even.